src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs

```diff
@@ -107,84 +124,137 @@ public async Task BlazorServerTemplateWorks_IndividualAuth(bool useLocalDB)
-                    TestBasicNavigation();
+                    EnsureBrowserAvailable(browserKind);
                 }
             }
+
+            void LogSocketTraffic(IWebSocket socket)
+            {
+                socket.FrameReceived += (s, e) => LogFrame((IWebSocket)s, sent: false, e.Payload);
+                socket.FrameSent += (s, e) => LogFrame((IWebSocket)s, sent: true, e.Payload);
+                socket.SocketError += (s, e) => LogSocketError((IWebSocket)s, e.ErrorMessage);
+                socket.Close += (s, e) => Output.WriteLine($"      [WebSocket][{((IWebSocket)s).Url}] is closed.");
+
+                void LogFrame(IWebSocket s, bool sent, string payload) =>
+                    Output.WriteLine($"      [WebSocket][{s.Url}] {(sent ? "sent" : "received")} frame with payload" +
+                    Environment.NewLine +
+                    $"      {payload}");
+
+                void LogSocketError(IWebSocket s, string message) =>
+                    Output.WriteLine($"      [WebSocket][{s.Url}] has error {message}");
+            }
+
+            void LogRequest(IRequest request)
+            {
+                Output.WriteLine($"      [Request]{(request.IsNavigationRequest ? "[Navigation]" : "")} {request.Method} {request.Url}");
+            }
+
+            void LogResponse(IResponse response)
+            {
+                var request = response.Request;
+                Output.WriteLine($"      [Response][{(int)response.Status}][{response.StatusText}]" + Environment.NewLine +
+                                 $"      For {(request.IsNavigationRequest ? "[Navigation]" : "")} {request.Method} {request.Url}");
+            }
         }
 
-        private void TestBasicNavigation()
+        private async Task TestBasicNavigation(IPage page)
         {
-            var retries = 3;
-            var connected = false;
-            do
-            {
-                try
-                {
-                    Browser.Contains("Information: WebSocket connected to",
-                        () => string.Join(Environment.NewLine, Browser.GetBrowserLogs(LogLevel.Info).Select(b => b.Message)));
-                    connected = true;
-                }
-                catch (TimeoutException) when(retries-- > 0)
-                {
-                    Browser.Navigate().Refresh();
-                }
-            } while (!connected && retries > 0);
+            var socket = (await page.WaitForEventAsync(PageEvent.WebSocket)).WebSocket;
+
+            // Start circuit
+            await socket.WaitForEventAsync(WebSocketEvent.FrameSent);
+
+            // Receive render batch
+            await socket.WaitForEventAsync(WebSocketEvent.FrameReceived);
+            await socket.WaitForEventAsync(WebSocketEvent.FrameSent);
+
+            // JS interop call to intercept navigation
+            await socket.WaitForEventAsync(WebSocketEvent.FrameReceived);
+            await socket.WaitForEventAsync(WebSocketEvent.FrameSent);
```

Comment:
I worry that this level of specificity is way too excessive for a test. We shouldn't be making assumptions about the protocol at such a fine-grained level. These details about websocket traffic counts and ordering aren't part of the behavioral contract and can be changed at any time. Especially hardcoding the `e.Payload == "ApEG"` is really disturbing.

Can we try to find a less implementation-specific way of describing behavior we're testing for?

Reply:
Ideally, could we replace everything to do with waiting for websocket traffic with a single "wait until the document gets changed in a certain way"? We can look for something that shows the prerendered content has been replaced by interactive content.

Reply:
I'll file a separate issue to follow up on this with the improvements we discussed offline.

Reply:
Link?

Reply:
https://github.com/dotnet/aspnetcore/issues/30116